File: takengine/takengine/src/main/cpp/renderer/raster/tilereader/GLTileMesh.cpp
#include "renderer/raster/tilereader/GLTileMesh.h"
#include "renderer/GLTexture2.h"
#include "renderer/GLES20FixedPipeline.h"
#include "renderer/core/GLMapView2.h"
#include "math/Utils.h"
#include "util/MathUtils.h"
#include "util/ConfigOptions.h"

using namespace TAK::Engine::Renderer::Raster;

namespace {
    int getConfigOption(const char *option, int defVal)
    {
        using namespace TAK::Engine;

        Port::String sval;
        Util::TAKErr err = Util::ConfigOptions_getOption(sval, option);
        if (err == Util::TE_InvalidArg) return defVal;
        int n;
        if (Port::String_parseInteger(&n, sval) == Util::TE_Ok) return n;
        return defVal;
    }
}


GLTileMesh::GLTileMesh(double width, double height, float u, float v, atakmap::raster::DatasetProjection *img2lla)
{
    commonInit(0, 0, width, height, 0.0f, 0.0f, u, v, img2lla);
}

GLTileMesh::GLTileMesh(double x, double y, double width, double height, float u0, float v0, float u1, float v1,
             atakmap::raster::DatasetProjection *img2lla) {
    commonInit(x, y, width, height, u0, v0, u1, v1, img2lla);
}

GLTileMesh::GLTileMesh(const TAK::Engine::Math::Point2<double> &imgUL, const TAK::Engine::Math::Point2<double> &imgUR,
                       const TAK::Engine::Math::Point2<double> &imgLR, const TAK::Engine::Math::Point2<double> &imgLL,
                       const TAK::Engine::Math::Matrix2 &img2uv,
                       atakmap::raster::DatasetProjection *img2lla,
                       int estimatedSubdivisions)
{
    commonInit(imgUL, imgUR, imgLR, imgLL, img2uv, img2lla, estimatedSubdivisions);
}

GLTileMesh::~GLTileMesh() { release(); }


TAK::Engine::Util::TAKErr GLTileMesh::commonInit(double x, double y, double width, double height, float u0, float v0, float u1, float v1,
                atakmap::raster::DatasetProjection *img2lla)
{
    Math::Matrix2 m;
    Math::Matrix2_mapQuads(&m, x, y, x + width, y, x + width, y + height, x, y + height, u0, v0, u1, v0, u1, v1, u0, v1);
    return commonInit(Math::Point2<double>(x, y), Math::Point2<double>(x + width, y), Math::Point2<double>(x + width, y + height),
               Math::Point2<double>(x, y + height), m, img2lla, estimateSubdivisions(x, y, width, height, img2lla));
}

    
TAK::Engine::Util::TAKErr GLTileMesh::commonInit(TAK::Engine::Math::Point2<double> imgUL, TAK::Engine::Math::Point2<double> imgUR,
                                                 TAK::Engine::Math::Point2<double> imgLR, TAK::Engine::Math::Point2<double> imgLL,
                                                 const TAK::Engine::Math::Matrix2 &img2uv, atakmap::raster::DatasetProjection *img2lla,
                                                 int estimatedSubdivisions) {
    mesh_coords_ = nullptr;
    mesh_coords_size_ = 0;
    mesh_verts_ = nullptr;
    mesh_verts_size_ = 0;
    mesh_tex_coords_ = nullptr;
    mesh_tex_coords_size_ = 0;
    mesh_indices_ = nullptr;
    mesh_indices_size_ = 0;
    num_coords_ = 0;
    num_indices_ = 0;
    mesh_verts_draw_version_ = 0;
    vert_mode_ = 0;
    mesh_dirty_ = false;
    this->img2lla_ = nullptr;
    this->estimated_subdivisions_ = 0;
    use_lcs_ = false;
    return resetMesh(imgUL, imgUR, imgLR, imgLL, img2uv, img2lla, estimatedSubdivisions);
}




TAK::Engine::Util::TAKErr GLTileMesh::resetMesh(double x, double y, double width, double height, float u0, float v0, float u1, float v1,
                           atakmap::raster::DatasetProjection *img2lla)
{
    Math::Matrix2 m;
    Math::Matrix2_mapQuads(&m, x, y, x + width, y, x + width, y + height, x, y + height, u0, v0, u1, v0, u1, v1, u0, v1);

    return resetMesh(Math::Point2<double>(x, y), Math::Point2<double>(x + width, y), Math::Point2<double>(x + width, y + height),
              Math::Point2<double>(x, y + height), m, img2lla,
                    estimateSubdivisions(x, y, width, height, img2lla));
}

TAK::Engine::Util::TAKErr GLTileMesh::resetMesh(TAK::Engine::Math::Point2<double> imgUL, TAK::Engine::Math::Point2<double> imgUR,
                                                TAK::Engine::Math::Point2<double> imgLR, TAK::Engine::Math::Point2<double> imgLL,
                                                TAK::Engine::Math::Matrix2 img2uv, atakmap::raster::DatasetProjection *img2lla,
                                                int estimatedSubdivisions) {
    this->img_ul_ = imgUL;
    this->img_ur_ = imgUR;
    this->img_lr_ = imgLR;
    this->img_ll_ = imgLL;
    this->img2uv_ = img2uv;
    TAK::Engine::Util::TAKErr err = this->img2uv_.createInverse(&this->uv2img_);
    if (err != TAK::Engine::Util::TE_Ok)
        return err;
    this->img2lla_ = img2lla;
    this->estimated_subdivisions_ = estimatedSubdivisions;

    // if it doesn't compute, that's ok
    this->centroid_ = TAK::Engine::Core::GeoPoint2(0, 0);
    atakmap::core::GeoPoint centroidgp(this->centroid_);
    this->use_lcs_ = this->img2lla_->imageToGround(
        atakmap::math::PointD((imgUL.x + imgUR.x + img_lr_.x + imgLL.x) / 4.0, (imgUL.y + imgUR.y + imgLR.y + imgLL.y) / 4.0, 0.0), &centroidgp);
    atakmap::core::GeoPoint_adapt(&this->centroid_, centroidgp);

    this->mesh_dirty_ = true;
    return TAK::Engine::Util::TE_Ok;
}

void GLTileMesh::validateMesh()
{
    if (!mesh_dirty_) return;

    num_indices_ = GLTexture2_getNumQuadMeshIndices(this->estimated_subdivisions_, this->estimated_subdivisions_);
    num_coords_ = GLTexture2_getNumQuadMeshVertices(this->estimated_subdivisions_, this->estimated_subdivisions_);

    if (this->mesh_tex_coords_ == nullptr || mesh_tex_coords_size_ < (num_coords_ * 2)) {
        if (mesh_tex_coords_)
            delete[] mesh_tex_coords_;

        mesh_tex_coords_size_ = num_coords_ * 2;
        mesh_tex_coords_ = new float[mesh_tex_coords_size_];
    }

    if (num_coords_ > 4) {
        if (mesh_indices_ == nullptr || mesh_indices_size_ < num_indices_) {
            if (mesh_indices_)
                delete[] mesh_indices_;

            mesh_indices_size_ = num_indices_;
            mesh_indices_ = new uint16_t[mesh_indices_size_];
        }
    } else if (mesh_indices_) {
        delete[] mesh_indices_;
        mesh_indices_ = nullptr;
        mesh_indices_size_ = 0;
    }

    if (mesh_verts_ == nullptr || mesh_verts_size_ < (num_coords_ * 3)) {
        if (mesh_verts_)
            delete[] mesh_verts_;

        mesh_verts_size_ = num_coords_ * 3;
        mesh_verts_ = new float[mesh_verts_size_];
    }

    Math::Point2<double> scratchD(0, 0, 0);

    img2uv_.transform(&scratchD, img_ul_);
    auto u0 = (float)scratchD.x;
    auto v0 = (float)scratchD.y;
    img2uv_.transform(&scratchD, img_ur_);
    auto u1 = (float)scratchD.x;
    auto v1 = (float)scratchD.y;
    img2uv_.transform(&scratchD, img_lr_);
    auto u2 = (float)scratchD.x;
    auto v2 = (float)scratchD.y;
    img2uv_.transform(&scratchD, img_ll_);
    auto u3 = (float)scratchD.x;
    auto v3 = (float)scratchD.y;

    GLTexture2_createQuadMeshTexCoords(mesh_tex_coords_, atakmap::math::Point<float>(u0, v0), atakmap::math::Point<float>(u1, v1),
                                       atakmap::math::Point<float>(u2, v2), atakmap::math::Point<float>(u3, v3),
                                        estimated_subdivisions_, estimated_subdivisions_);

    if (mesh_indices_ != nullptr) {
        GLTexture2_createQuadMeshIndexBuffer(mesh_indices_, estimated_subdivisions_, estimated_subdivisions_);
    }

    vert_mode_ = GL_TRIANGLE_STRIP;

    // XXX - generate the LLA coords from the texcoords
    if (mesh_coords_ == nullptr || mesh_coords_size_ < (num_coords_ * 2)) {
        if (mesh_coords_)
            delete[] mesh_coords_;
        mesh_coords_size_ = num_coords_ * 2;
        mesh_coords_ = new double[mesh_coords_size_];
    }

    scratchD.z = 0;

    atakmap::core::GeoPoint scratchGeo;

    float *meshTexCoordsPtr = mesh_tex_coords_;
    double *meshCoordsPtr = mesh_coords_;
    for (std::size_t i = 0u; i < num_coords_; i++) {
        scratchD.x = *meshTexCoordsPtr;
        meshTexCoordsPtr++;
        scratchD.y = *meshTexCoordsPtr;
        meshTexCoordsPtr++;

        // XXX - uv2img transform pointD
        uv2img_.transform(&scratchD, scratchD);

        img2lla_->imageToGround(atakmap::math::PointD(scratchD.x, scratchD.y, scratchD.z), &scratchGeo);
        *meshCoordsPtr = scratchGeo.longitude;
        meshCoordsPtr++;
        *meshCoordsPtr = scratchGeo.latitude;
        meshCoordsPtr++;
    }

    mesh_verts_draw_version_ = -1;
    mesh_dirty_ = false;
}

void GLTileMesh::setLCS(const TAK::Engine::Renderer::Core::GLGlobeBase &view, bool primary) {
    Math::Matrix2 scratchM;
    scratchM.setToIdentity();
    scratchM.concatenate(view.renderPass->scene.forwardTransform);

    Math::Point2<double> pointD;
    TAK::Engine::Core::GeoPoint2 geo;

    if (!primary) {
        if (centroid_.longitude >= 0.0)
            geo = TAK::Engine::Core::GeoPoint2(centroid_.latitude, centroid_.longitude - 360.0, 0.0, TAK::Engine::Core::AltitudeReference::HAE);
        else
            geo = TAK::Engine::Core::GeoPoint2(centroid_.latitude, centroid_.longitude + 360.0, 0.0, TAK::Engine::Core::AltitudeReference::HAE);
        view.renderPass->scene.projection->forward(&pointD, geo);

        scratchM.translate(pointD.x, pointD.y, pointD.z);
    } else {
        scratchM.translate(centroid_proj_.x, centroid_proj_.y, centroid_proj_.z);
    }

    double scratchD[16];
    float scratchF[16];
    scratchM.get(scratchD, Math::Matrix2::COLUMN_MAJOR);
    for (int i = 0; i < 16; i++)
        scratchF[i] = (float)scratchD[i];
    atakmap::renderer::GLES20FixedPipeline *fixedPipe = atakmap::renderer::GLES20FixedPipeline::getInstance();
    fixedPipe->glLoadMatrixf(scratchF);
}

void GLTileMesh::drawMesh(const TAK::Engine::Renderer::Core::GLGlobeBase &view, int texId, float r, float g, float b, float a) {
    validateMesh();

    Math::Point2<double> pointD;
    TAK::Engine::Core::GeoPoint2 geo;

    if (mesh_verts_draw_version_ != view.renderPass->drawSrid) {
        if (use_lcs_) {
            view.renderPass->scene.projection->forward(&centroid_proj_, centroid_);
        } else {
            centroid_proj_.x = 0;
            centroid_proj_.y = 0;
            centroid_proj_.z = 0;
        }

        double *meshCoordsPtr = mesh_coords_;
        float *meshVertsPtr = mesh_verts_;
        for (std::size_t i = 0u; i < num_coords_; i++) {
            double lon = *meshCoordsPtr;
            meshCoordsPtr++;
            double lat = *meshCoordsPtr;
            meshCoordsPtr++;
            geo = TAK::Engine::Core::GeoPoint2(lat, lon, 0.0, TAK::Engine::Core::AltitudeReference::HAE);

            view.renderPass->scene.projection->forward(&pointD, geo);
            *meshVertsPtr = (float)(pointD.x - centroid_proj_.x);
            meshVertsPtr++;
            *meshVertsPtr = (float)(pointD.y - centroid_proj_.y);
            meshVertsPtr++;
            *meshVertsPtr = (float)(pointD.z - centroid_proj_.z);
            meshVertsPtr++;
        }

        mesh_verts_draw_version_ = view.renderPass->drawSrid;
    }

    atakmap::renderer::GLES20FixedPipeline *fixedPipe = atakmap::renderer::GLES20FixedPipeline::getInstance();
    fixedPipe->glPushMatrix();

    const int pumps = view.renderPass->crossesIDL ? 2 : 1;
    for (int i = 0; i < pumps; i++) {
        setLCS(view, i % 2 == 0);

        if (mesh_indices_ != nullptr) {
            GLTexture2_draw(texId, vert_mode_, num_indices_, 2, GL_FLOAT, mesh_tex_coords_, 3, GL_FLOAT, mesh_verts_, GL_UNSIGNED_SHORT,
                            mesh_indices_, r, g, b, a);
        } else {
            GLTexture2_draw(texId, vert_mode_, num_coords_, 2, GL_FLOAT, mesh_tex_coords_, 3, GL_FLOAT, mesh_verts_, r, g, b, a);
        }
    }

    fixedPipe->glPopMatrix();
}

void GLTileMesh::release() {
    if (mesh_coords_ != nullptr) {
        delete[] mesh_coords_;
        mesh_coords_ = nullptr;
        mesh_coords_size_ = 0;
    }
    if (mesh_verts_ != nullptr) {
        delete[] mesh_verts_;
        mesh_verts_ = nullptr;
        mesh_verts_size_ = 0;
    }
    if (mesh_tex_coords_ != nullptr) {
        delete[] mesh_tex_coords_;
        mesh_tex_coords_ = nullptr;
        mesh_tex_coords_size_ = 0;
    }
    if (mesh_indices_ != nullptr) {
        delete[] mesh_indices_;
        mesh_indices_ = nullptr;
        mesh_indices_size_ = 0;
    }
    num_coords_ = 0;
    num_indices_ = 0;

    mesh_verts_draw_version_ = -1;

    mesh_dirty_ = true;
}

int GLTileMesh::estimateSubdivisions(double ulLat, double ulLng, double lrLat, double lrLng) {
    const int minGridSize = getConfigOption("glquadtilenode2.minimum-grid-size", 1);
    const int maxGridSize = getConfigOption("glquadtilenode2.maximum-grid-size", 32);

    const int subsX = Util::MathUtils_clamp(
        Util::MathUtils_nextPowerOf2((int)ceil((ulLat - lrLat) / TAK::Engine::Renderer::Core::GLMapView2::getRecommendedGridSampleDistance())),
                                        minGridSize, maxGridSize);
    const int subsY = Util::MathUtils_clamp(
        Util::MathUtils_nextPowerOf2((int)ceil((lrLng - ulLng) / TAK::Engine::Renderer::Core::GLMapView2::getRecommendedGridSampleDistance())),
                                        minGridSize, maxGridSize);

    return atakmap::math::max(subsX, subsY);
}

int GLTileMesh::estimateSubdivisions(double x, double y, double width, double height, atakmap::raster::DatasetProjection *img2lla) {
    double minLat;
    double minLng;
    double maxLat;
    double maxLng;

    atakmap::math::PointD p = atakmap::math::PointD(0.0, 0.0);
    atakmap::core::GeoPoint g;

    p.x = x;
    p.y = y;
    img2lla->imageToGround(p, &g);
    minLat = g.latitude;
    minLng = g.longitude;
    maxLat = g.latitude;
    maxLng = g.longitude;

    p.x = x + width;
    p.y = y;
    img2lla->imageToGround(p, &g);
    minLat = atakmap::math::min(g.latitude, minLat);
    minLng = atakmap::math::min(g.longitude, minLng);
    maxLat = atakmap::math::max(g.latitude, maxLat);
    maxLng = atakmap::math::max(g.longitude, maxLng);

    p.x = x + width;
    p.y = y + height;
    img2lla->imageToGround(p, &g);
    minLat = atakmap::math::min(g.latitude, minLat);
    minLng = atakmap::math::min(g.longitude, minLng);
    maxLat = atakmap::math::max(g.latitude, maxLat);
    maxLng = atakmap::math::max(g.longitude, maxLng);

    p.x = x;
    p.y = y + height;
    img2lla->imageToGround(p, &g);
    minLat = atakmap::math::min(g.latitude, minLat);
    minLng = atakmap::math::min(g.longitude, minLng);
    maxLat = atakmap::math::max(g.latitude, maxLat);
    maxLng = atakmap::math::max(g.longitude, maxLng);


    const int minGridSize = getConfigOption("glquadtilenode2.minimum-grid-size", 1);
    const int maxGridSize = getConfigOption("glquadtilenode2.maximum-grid-size", 32);

    const int subsX = Util::MathUtils_clamp(
        Util::MathUtils_nextPowerOf2((int)ceil((maxLat - minLat) / TAK::Engine::Renderer::Core::GLMapView2::getRecommendedGridSampleDistance())),
        minGridSize, maxGridSize);
    const int subsY = Util::MathUtils_clamp(
        Util::MathUtils_nextPowerOf2((int)ceil((maxLng - minLng) / TAK::Engine::Renderer::Core::GLMapView2::getRecommendedGridSampleDistance())),
        minGridSize, maxGridSize);

    return atakmap::math::max(subsX, subsY);
}

File: takkernel/engine/src/main/java/com/atakmap/map/layer/raster/tilereader/opengl/GLTileMesh.java
package com.atakmap.map.layer.raster.tilereader.opengl;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.DoubleBuffer;
import java.nio.FloatBuffer;

import android.graphics.PointF;


import com.atakmap.coremap.maps.coords.GeoPoint;
import com.atakmap.lang.Unsafe;
import com.atakmap.map.layer.raster.DatasetProjection2;
import com.atakmap.map.opengl.GLMapView;
import com.atakmap.math.MathUtils;
import com.atakmap.math.Matrix;
import com.atakmap.math.NoninvertibleTransformException;
import com.atakmap.math.PointD;
import com.atakmap.opengl.GLES20FixedPipeline;
import com.atakmap.opengl.GLTexture;
import com.atakmap.util.ConfigOptions;
import com.atakmap.util.Releasable;

public class GLTileMesh implements Releasable {


    private DoubleBuffer meshCoords;
    private FloatBuffer meshVerts;
    private FloatBuffer meshTexCoords;
    private ByteBuffer meshIndices;
    private int numCoords;
    private int numIndices;
    private final int meshVertsSize;
    private int meshVertsDrawVersion;
    private int vertMode;
    private boolean meshDirty;

    private DatasetProjection2 img2lla;
    private Matrix img2uv;
    private Matrix uv2img;
    private PointD imgUL;
    private PointD imgUR;
    private PointD imgLR;
    private PointD imgLL;
    private int estimatedSubdivisions;

    private GeoPoint centroid;
    private PointD centroidProj;
    private boolean useLCS;
    
    public GLTileMesh(double width,
                      double height,
                      float u,
                      float v,
                      DatasetProjection2 img2lla) {
        
        this(0, 0, width, height, 0f, 0f, u, v, img2lla);
    }
    
    public GLTileMesh(double x,
                      double y,
                      double width,
                      double height,
                      float u0,
                      float v0,
                      float u1,
                      float v1,
                      DatasetProjection2 img2lla) {

        this(new PointD(x, y),
             new PointD(x+width, y),
             new PointD(x+width, y+height),
             new PointD(x, y+height),
             Matrix.mapQuads(x, y,
                             x+width, y,
                             x+width, y+height,
                             x, y+height,
                             u0, v0,
                             u1, v0,
                             u1, v1,
                             u0, v1),
             img2lla,
             estimateSubdivisions(x, y, width, height, img2lla));
    }

    public GLTileMesh(PointD imgUL,
                      PointD imgUR,
                      PointD imgLR,
                      PointD imgLL,
                      Matrix img2uv,
                      DatasetProjection2 img2lla,
                      int estimatedSubdivisions) {

        this.meshVertsSize = 3;
        this.centroid = GeoPoint.createMutable();
        this.centroidProj = new PointD(0d, 0d, 0d);
        this.useLCS = false;
        this.resetMesh(imgUL,
                       imgUR,
                       imgLR,
                       imgLL,
                       img2uv,
                       img2lla,
                       estimatedSubdivisions);
    }

    public void resetMesh(double x,
                          double y,
                          double width,
                          double height,
                          float u0,
                          float v0,
                          float u1,
                          float v1,
                          DatasetProjection2 img2lla) {

        this.resetMesh(new PointD(x, y),
                      new PointD(x+width, y),
                      new PointD(x+width, y+height),
                      new PointD(x, y+height),
                      Matrix.mapQuads(x, y,
                                      x+width, y,
                                      x+width, y+height,
                                      x, y+height,
                                      u0, v0,
                                      u1, v0,
                                      u1, v1,
                                      u0, v1),
                      img2lla,
                      estimateSubdivisions(x, y, width, height, img2lla));
    }
    
    public void resetMesh(PointD imgUL,
                          PointD imgUR,
                          PointD imgLR,
                          PointD imgLL,
                          Matrix img2uv,
                          DatasetProjection2 img2lla,
                          int estimatedSubdivisions) {
        
        this.imgUL = imgUL;
        this.imgUR = imgUR;
        this.imgLR = imgLR;
        this.imgLL = imgLL;
        this.img2uv = img2uv;
        try {
            this.uv2img = this.img2uv.createInverse();
        } catch(NoninvertibleTransformException e) {
            throw new IllegalArgumentException();
        }
        this.img2lla = img2lla;
        this.estimatedSubdivisions = estimatedSubdivisions;

        // if it doesn't compute, that's ok
        this.centroid.set(0d, 0d);
        this.useLCS = this.img2lla.imageToGround(
                new PointD((imgUL.x+imgUR.x+imgLR.x+imgLL.x)/4d,
                           (imgUL.y+imgUR.y+imgLR.y+imgLL.y)/4d,
                        0d),
                this.centroid);
        
        this.meshDirty = true;
    }

    private void validateMesh() {
        if(!this.meshDirty)
            return;
        
        this.numIndices = GLTexture.getNumQuadMeshIndices(this.estimatedSubdivisions,
                                                          this.estimatedSubdivisions);
        this.numCoords = GLTexture.getNumQuadMeshVertices(this.estimatedSubdivisions,
                                                          this.estimatedSubdivisions);

        if (this.meshTexCoords == null
                || this.meshTexCoords.capacity() < (this.numCoords * 2)) {
            
            Unsafe.free(this.meshTexCoords);

            ByteBuffer buf = Unsafe.allocateDirect(this.numCoords * 8);
            buf.order(ByteOrder.nativeOrder());
            
            this.meshTexCoords = buf.asFloatBuffer();
        }

        if (this.numCoords > 4) {
            if (this.meshIndices == null
                    || this.meshIndices.capacity() < (this.numIndices * 2)) {
                
                Unsafe.free(this.meshIndices);

                this.meshIndices = Unsafe.allocateDirect(this.numIndices * 2);
                this.meshIndices.order(ByteOrder.nativeOrder());
            }
        } else {
            Unsafe.free(this.meshIndices);
            this.meshIndices = null;
        }

        if (this.meshVerts == null
                || this.meshVerts.capacity() < (this.numCoords * 3)) {
            
            Unsafe.free(this.meshVerts);
            this.meshVerts = Unsafe.allocateDirect(this.numCoords*3, FloatBuffer.class);
        } else {
            this.meshVerts.clear();
        }

        PointD scratchD = new PointD(0d, 0d, 0d);
        
        img2uv.transform(this.imgUL, scratchD);
        float u0 = (float)scratchD.x;
        float v0 = (float)scratchD.y;
        img2uv.transform(this.imgUR, scratchD);
        float u1 = (float)scratchD.x;
        float v1 = (float)scratchD.y;
        img2uv.transform(this.imgLR, scratchD);
        float u2 = (float)scratchD.x;
        float v2 = (float)scratchD.y;
        img2uv.transform(this.imgLL, scratchD);
        float u3 = (float)scratchD.x;
        float v3 = (float)scratchD.y;
        
        this.meshTexCoords.clear();
        GLTexture.createQuadMeshTexCoords(new PointF(u0, v0),
                                          new PointF(u1, v1),
                                          new PointF(u2, v2),
                                          new PointF(u3, v3),
                                          this.estimatedSubdivisions,
                                          this.estimatedSubdivisions,
                                          this.meshTexCoords);
        this.meshTexCoords.flip();

        if (this.meshIndices != null) {
            GLTexture.createQuadMeshIndexBuffer(
                    this.estimatedSubdivisions,
                    this.estimatedSubdivisions,
                    this.meshIndices.asShortBuffer());
        }
        
        this.vertMode = GLES20FixedPipeline.GL_TRIANGLE_STRIP;
        
        // XXX - generate the LLA coords from the texcoords
        if(this.meshCoords == null || (this.meshCoords.capacity() < (this.numCoords*2))) {
            Unsafe.free(this.meshCoords);
            this.meshCoords = Unsafe.allocateDirect(this.numCoords*2, DoubleBuffer.class);
        }

        scratchD.z = 0d;

        GeoPoint scratchGeo = GeoPoint.createMutable();

        long meshTexCoordsPtr = Unsafe.getBufferPointer(this.meshTexCoords);
        long meshCoordsPtr = Unsafe.getBufferPointer(this.meshCoords);
        for(int i = 0; i < this.numCoords; i++) {
            scratchD.x = Unsafe.getFloat(meshTexCoordsPtr);
            scratchD.y = Unsafe.getFloat(meshTexCoordsPtr+4);
            meshTexCoordsPtr += 8;
            
            // XXX - uv2img transform pointD
            this.uv2img.transform(scratchD, scratchD);
            
            this.img2lla.imageToGround(scratchD, scratchGeo);
            Unsafe.setDoubles(meshCoordsPtr, scratchGeo.getLongitude(), scratchGeo.getLatitude());
            meshCoordsPtr += 16;
        }

        this.meshVertsDrawVersion = -1;
        this.meshDirty = false;
    }

    private void setLCS(GLMapView view, boolean primary) {
        view.scratch.matrix.setToIdentity();
        view.scratch.matrix.concatenate(view.scene.forward);
        if(!primary) {
            if(this.centroid.getLongitude() >= 0d)
                view.scratch.geo.set(this.centroid.getLatitude(), this.centroid.getLongitude()-360d, 0d);
            else
                view.scratch.geo.set(this.centroid.getLatitude(), this.centroid.getLongitude()+360d, 0d);
            view.scene.mapProjection.forward(view.scratch.geo, view.scratch.pointD);

            view.scratch.matrix.translate(view.scratch.pointD.x, view.scratch.pointD.y, view.scratch.pointD.z);
        } else {
            view.scratch.matrix.translate(centroidProj.x, centroidProj.y, centroidProj.z);
        }

        view.scratch.matrix.get(view.scratch.matrixD, Matrix.MatrixOrder.COLUMN_MAJOR);
        for(int i = 0; i < 16; i++)
            view.scratch.matrixF[i] = (float)view.scratch.matrixD[i];
        GLES20FixedPipeline.glLoadMatrixf(view.scratch.matrixF, 0);
    }

    public void drawMesh(GLMapView view, int texId, float r, float g, float b, float a) {
        this.validateMesh();

        if(this.meshVertsDrawVersion != view.drawSrid) {
            view.scratch.geo.set(GeoPoint.UNKNOWN);
            if(this.useLCS) {
                view.scene.mapProjection.forward(this.centroid, this.centroidProj);
            } else {
                this.centroidProj.x = 0d;
                this.centroidProj.y = 0d;
                this.centroidProj.z = 0d;
            }

            long meshCoordsPtr = Unsafe.getBufferPointer(this.meshCoords);
            long meshVertsPtr = Unsafe.getBufferPointer(this.meshVerts);
            for(int i = 0; i < this.numCoords; i++) {
                view.scratch.geo.set(Unsafe.getDouble(meshCoordsPtr+8),
                                     Unsafe.getDouble(meshCoordsPtr),
                                     0d);
                meshCoordsPtr += 16;
                
                view.scene.mapProjection.forward(view.scratch.geo, view.scratch.pointD);
                Unsafe.setFloats(meshVertsPtr,
                                 (float)(view.scratch.pointD.x - this.centroidProj.x),
                                 (float)(view.scratch.pointD.y - this.centroidProj.y),
                                 (float)(view.scratch.pointD.z - this.centroidProj.z));
                meshVertsPtr += 12;
            }

            this.meshVertsDrawVersion = view.drawSrid;
        }

        GLES20FixedPipeline.glPushMatrix( );

        setLCS(view, (view.currentPass.drawLng*centroid.getLongitude()) >= 0 || view.drawSrid != 4326);

        if (meshIndices != null) {
            GLTexture.draw(texId,
                    vertMode,
                    numIndices,
                    2, GLES20FixedPipeline.GL_FLOAT, meshTexCoords,
                    meshVertsSize, GLES20FixedPipeline.GL_FLOAT, meshVerts,
                    GLES20FixedPipeline.GL_UNSIGNED_SHORT, meshIndices,
                    r, g, b, a);
        } else {
            GLTexture.draw(texId,
                    vertMode,
                    numCoords,
                    2, GLES20FixedPipeline.GL_FLOAT, meshTexCoords,
                    meshVertsSize, GLES20FixedPipeline.GL_FLOAT, meshVerts,
                    r, g, b, a);
        }

        GLES20FixedPipeline.glPopMatrix();
    }

    @Override
    public void release() {
        if(this.meshCoords != null) {
            Unsafe.free(this.meshCoords);
            this.meshCoords = null;
        }
        if(this.meshVerts != null) {
            Unsafe.free(this.meshVerts);
            this.meshVerts = null;
        }
        if(this.meshTexCoords != null) {
            Unsafe.free(this.meshTexCoords);
            this.meshTexCoords = null;
        }
        if(this.meshIndices != null) {
            Unsafe.free(this.meshIndices);
            this.meshIndices = null;
        }
        this.numCoords = 0;
        this.numIndices = 0;

        this.meshVertsDrawVersion = -1;
        
        this.meshDirty = true;
    }

    public static int estimateSubdivisions(double ulLat, double ulLng, double lrLat, double lrLng) {
        final int minGridSize = ConfigOptions.getOption("glquadtilenode2.minimum-grid-size", 1);
        final int maxGridSize = ConfigOptions.getOption("glquadtilenode2.maximum-grid-size", 32);

        final int subsX = MathUtils.clamp(MathUtils.nextPowerOf2((int)Math.ceil((ulLat-lrLat) / GLMapView.recommendedGridSampleDistance)), minGridSize, maxGridSize);
        final int subsY = MathUtils.clamp(MathUtils.nextPowerOf2((int)Math.ceil((lrLng-ulLng) / GLMapView.recommendedGridSampleDistance)), minGridSize, maxGridSize);
        
        return Math.max(subsX, subsY);
    }
    
    private static int estimateSubdivisions(double x, double y, double width, double height, DatasetProjection2 img2lla) {
        double minLat;
        double minLng;
        double maxLat;
        double maxLng;
        
        PointD p = new PointD(0d, 0d);
        GeoPoint g = GeoPoint.createMutable();
        
        p.x = x;
        p.y = y;
        img2lla.imageToGround(p, g);
        minLat = g.getLatitude();
        minLng = g.getLongitude();
        maxLat = g.getLatitude();
        maxLng = g.getLongitude();
        
        p.x = x + width;
        p.y = y;
        img2lla.imageToGround(p, g);
        minLat = Math.min(g.getLatitude(), minLat);
        minLng = Math.min(g.getLongitude(), minLng);
        maxLat = Math.max(g.getLatitude(), maxLat);
        maxLng = Math.max(g.getLongitude(), maxLng);
        
        p.x = x + width;
        p.y = y + height;
        img2lla.imageToGround(p, g);
        minLat = Math.min(g.getLatitude(), minLat);
        minLng = Math.min(g.getLongitude(), minLng);
        maxLat = Math.max(g.getLatitude(), maxLat);
        maxLng = Math.max(g.getLongitude(), maxLng);
        
        p.x = x;
        p.y = y + height;
        img2lla.imageToGround(p, g);
        minLat = Math.min(g.getLatitude(), minLat);
        minLng = Math.min(g.getLongitude(), minLng);
        maxLat = Math.max(g.getLatitude(), maxLat);
        maxLng = Math.max(g.getLongitude(), maxLng);
        
        final int minGridSize = ConfigOptions.getOption("glquadtilenode2.minimum-grid-size", 1);
        final int maxGridSize = ConfigOptions.getOption("glquadtilenode2.maximum-grid-size", 32);

        final int subsX = MathUtils.clamp(MathUtils.nextPowerOf2((int)Math.ceil((maxLat-minLat) / GLMapView.recommendedGridSampleDistance)), minGridSize, maxGridSize);
        final int subsY = MathUtils.clamp(MathUtils.nextPowerOf2((int)Math.ceil((maxLng-minLng) / GLMapView.recommendedGridSampleDistance)), minGridSize, maxGridSize);
        
        return Math.max(subsX, subsY);
    }
}
